### DISCUSSION TITLE
Question on Delay by create_voice_channel

### DISCUSSION THREAD

**Discussion by Kartoffel096** [Rapptz/discord.py#9755] [Q&A]
Hello there, i am pretty new to writing Discord Bots and just installed and used this library to make a Bot that Creates User Channels when joining a predefined Channel. 
This works just fine but it sometimes has like a 4-5 second delay and i wan't to know why. 

```
async def on_channel_join(member, after, _SERVER, _category):
    max_users = after.channel.user_limit

    if after.channel.id in cmd_channel:
        name = f"{member.name}'s Channel"

        channel = await _SERVER.create_voice_channel(
            name, user_limit=max_users, category=_category
        )

        user_channels.append(channel)
        await member.move_to(channel)
    return
```
    
  That is my Code and with lot's of testing and print statements i could pin down the delay to this one line
  `channel = await _SERVER.create_voice_channel(
            name, user_limit=max_users, category=_category
        )`
        
   It seems like this call sometimes takes 4-5s and sometimes just milliseconds
   
   Is there any Reasons why this takes so long to process or is there anything i can do or provide you with?

**Comment by LostLuma** [comment]
The API has rate limits in place to prevent abuse which control how often your bot can do certain actions, meaning the reason you're sometimes experiencing a delay of a few seconds is probably because you've recently created another channel in the same server.

Note that there is AFAIK a limit on how many channels you can create / edit / delete per day as well as the limit you already in to, so this feature of yours might not work with larger or busier communities.

**Reply by Kartoffel096** [reply]
I thought about rate limiting too but fortunatly there is a nice Feature to test that `client.is_ws_ratelimited()` which returns false so i guess i am not getting rate limited by the discord api. I also checked the websocket latency with `client.latency` which is always around 0.1 seconds.

The Server is a Communityserver with about 1300 Users from which about a third is active.

Do you have any other Ideas why this could be slow?

**Reply by LostLuma** [reply]
Whether the websocket is rate limited doesn't really matter here as creating channels uses the HTTP API.

What you could do to confirm whether my assumption above is correct or there is another cause is to enable debug logging for the `discord.http` logger and then recreate the situation in which it is slow. You can do so using the built-in logging library:

`logging.getLogger('discord.http').setLevel(logging.DEBUG)`

Note that this will also create a lot of other irrelevant logs, so you might want to disable other features of your bot temporarily.

**Reply by Kartoffel096** [reply]
I used the Advanced Logging now with the discord.http and there was no rate limiting the request returned a 201 but the timestamp of the POST request with the 201 return was about 4 seconds late. So i guess maybe the asyncio on that Server takes about 4 seconds as overhead to create the thread and send the request ? Could that be possible? I think with this tests we can rule out the API itself. 
The problem unfortunatly still persists and i have no idea why sadly.

**Comment by Kartoffel096** [comment]
Could it be that the delay is dependend on the Size of the Server ?
I testet the Bot on a small local testing Server with just 3 Members and a handfull of channels where it was snipping fast. But as soon as it is on our Community Server with 1000+ Members and a bit more Channels it slows down.

Shouldn't the Bot by default cache Members and Channels ? Could that be even the cause for the slow ?

**Reply by LostLuma** [reply]
> Could it be that the delay is dependent on the Size of the Server ?

I tested this for you and it doesn't seem server size (channel or member count, whichever you mean) has any effect whatsoever.

> Shouldn't the Bot by default cache Members and Channels ? Could that be even the cause for the slow ?

Whether the bot caches members or channels is also not relevant, all `Guild.create_voice_channel` does internally is call the [Create Guild Channel](https://discord.com/developers/docs/resources/guild#create-guild-channel) endpoint on your behalf with the parameters provided, and then adds the newly created channel to the Guild's internal cache (without caring about whether other channels are cached or not, it's just added to it).

---

What might be interesting to know is whether you are testing in the small and larger server using the same bot account and are running it on the exact same computer, or is it separate bots and computers? It could be as simple as your production computer having a worse internet connection (do commands or other actions the bot does take the same amount of time?), or just being too low-power (if it's a Raspberry PI or something similar?) to run it. In the same vein maybe also post which intents you are using.

**Reply by Kartoffel096** [reply]
> What might be interesting to know is whether you are testing in the small and larger server using the same bot account and are running it on the exact same computer, or is it separate bots and computers? It could be as simple as your production computer having a worse internet connection (do commands or other actions the bot does take the same amount of time?), or just being too low-power (if it's a Raspberry PI or something similar?) to run it. In the same vein maybe also post which intents you are using.

I am using the same Bot Account for testing and for production. For testing i use my local Machine(i7-7700K; 64GB RAM), for production i am using a Docker Container (4 CPU Cores, 2GiB of Memory) on a Server(i9-9900K; 128GB RAM) which hosts some other Containers as well. I personally have a 1Gbit Internet Connection as well as the Server.

I am currently using `discord.intents.all()` for testing and i am looking to use `discord.intents(emojis=True,guild_messages=True,guild_reactions=True,members=True,messages=True,moderation=True,reactions=True,voice_states=True)` for production later on

**Comment by Kartoffel096** [comment]
I solved the Problem by using Bare-Metal Hardware and running the Bot there instead of inside a Docker Container. Seems like there are some problemes with running Python inside a container regarding the performance.

Thank you for taking the Time and helping me out.
